lib/mongoid/matcher.rb

```diff
@@ -47,9 +47,8 @@ module Matcher
         current.each do |doc|
           case doc
           when Hash
-            v = indifferent_hash_fetch(doc, field)
-            if !v.nil?
-              new << v
+            if indifferent_key?(doc, field)
```

Comment:
How about something like this instead?

```ruby
def indifferent_fetch(h, k)
  if h.key?(k.to_s)
    yield h[k.to_s]
  elsif h.key?(k.to_sym)
    yield h[k.to_sym]
  end
end

h1 = { a: 1, b: nil }
h2 = {}

indifferent_fetch(h1, "a") do |v|
  h2[:a2] = v
end

indifferent_fetch(h1, :b) do |v|
  h2[:b2] = v
end

indifferent_fetch(h1, "c") do |v|
  h2[:c2] = v
end

print("h2:\n#{h2}\n")
# {:a2=>1, :b2=>nil}
```

I believe would end up with fewer key accesses.